### ISSUE TITLE
Error when running inspect and using horizontal wrappers

### ISSUE THREAD

**Issue by kinow** [BSC-ES/autosubmit#1259]
In GitLab by @Lerriola on Mar 21, 2024, 08:58

Hello @dbeltrankyl and @kinow,

### Autosubmit Version

v4.1.2

### Expid affected(If applicable)

The experiment is **a0yd** the one I was using for the ICON transfer jobs. The experiment didn't have wrappers before, and I tried to add them afterwards.

### Which task has issues? Where is the log(If applicable)

* **Full_name**: a0yd_19910301_fc0_{X}_TRANSFER

* **Log_Path:** 
### Summary

I'm not sure if it's only one bug or there are more layers to it. The bug or weird Autosubmit behaviour first appears when running inspect. After taking a long while, which I don't know if it's usual (I'm running a job with 1828 jobs though) Autosubmit crashes and reports the following [error.txt](https://earth.bsc.es/gitlab/es/autosubmit/uploads/713c5cebe77ab43c57dcdaa8fc8285d6/error.txt) This is part of the error that appears.

The thing is, it breaks Autosubmit and then doesn't show much in the `ASLOGS/20240321_093513_inspect.log` logs.

After running inspect, you can see that the files are actually still produced. Yesterday I tried running the experiment, but what happened was that Autosubmit failed to be able to submit the wrapper jobs, and as it was set to “flexible” it directly just submitted all the jobs it could...

### Steps to reproduce

- access the climateDT VM (using the shared account)
- load module autosubmit/4.1.2
- autosubmit inspect a0yd
- autosubmit run a0yd

### What is the current bug behavior?

It breaks and doesn't allow to run using the horizontal wrappers.

### What is the expected correct behavior?

It creates the jobs well during inspect, and then it can submit wrappers taking into account the already completed jobs, etc...

### Relevant logs and/or screenshots(if applicable)

[error.txt](https://earth.bsc.es/gitlab/es/autosubmit/uploads/713c5cebe77ab43c57dcdaa8fc8285d6/error.txt)

### Any other relevant information(if applicable)

()

**Comment by kinow** [comment]
In GitLab by @Lerriola on Mar 21, 2024, 08:58

created branch [`1259-error-when-running-inspect-and-usnig-horizontal-wrappers`](/gitlab/es/autosubmit/-/compare/master...1259-error-when-running-inspect-and-usnig-horizontal-wrappers) to address this issue

**Comment by kinow** [comment]
In GitLab by @Lerriola on Mar 21, 2024, 08:59

I accidentally created the branch sorry!

**Comment by kinow** [comment]
In GitLab by @Lerriola on Mar 21, 2024, 09:03

mentioned in issue digital-twins/de_340/workflow#259

**Comment by kinow** [comment]
In GitLab by @Lerriola on Mar 21, 2024, 09:06

Also, which might give a bit more insight, is that when trying to run with another version (p.e 4.0.104) I get the following when running inspect:


```
During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/appl/AS/4.0.104/lib/python3.9/site-packages/autosubmit-4.0.104-py3.9.egg/EGG-INFO/scripts/autosubmit", line 37, in main
    return_value = Autosubmit.parse_args()
  File "/appl/AS/4.0.104/lib/python3.9/site-packages/autosubmit-4.0.104-py3.9.egg/autosubmit/autosubmit.py", line 684, in parse_args
    return Autosubmit.inspect(args.expid, args.list, args.filter_chunks, args.filter_status,
  File "/appl/AS/4.0.104/lib/python3.9/site-packages/autosubmit-4.0.104-py3.9.egg/autosubmit/autosubmit.py", line 1537, in inspect
    raise AutosubmitCritical(
log.log.AutosubmitCritical:  
[ERROR] Trace: 'int' object has no attribute 'get'
 [CRITICAL] There are issues that occurred during the templates generation, please check that job parameters are well set and the template path exists. [eCode=7014]
```

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Mar 21, 2024, 09:38

Hello @Lerriola, 

I'm a bit confused. Which command failed to run? 

> but what happened was that Autosubmit failed to be able to submite the wrapper jobs, and as it was set to “flexible” it directly just submitted all the jobs it could

What do you mean by failing to submit the wrapper job? 

What Autosubmit does with the flexible policy is to submit wrappers as they come. If they don't fit the min,max wrapped introduced they will be sent as single jobs. 

To avoid this, you can use the mixed or strict policy 

Also, to inspect the wrappers cmds you should add the flag -cw

**Comment by kinow** [comment]
In GitLab by @Lerriola on Mar 21, 2024, 09:42

Hi @dbeltrankyl,
First the inspect command and then when running it seemed as it hadn't submitted the wrapper.

**Comment by kinow** [comment]
In GitLab by @Lerriola on Mar 21, 2024, 10:37

The error which I now get when doing AS run, which think is simular to the one in the inspect is the following:

```
Exception in thread data_processing:
Traceback (most recent call last):
  File "/usr/lib64/python3.9/threading.py", line 980, in _bootstrap_inner
    self.run()
  File "/usr/lib64/python3.9/threading.py", line 917, in run
    self._target(*self._args, **self._kwargs)
  File "/appl/AS/4.1.2/lib/python3.9/site-packages/autosubmit-4.1.2-py3.9.egg/autosubmit/job/job_packages.py", line 103, in check_scripts
    raise AutosubmitCritical(f"Job script:{job.file} does not exists",7014)
log.log.AutosubmitCritical:  
Exception in thread data_processing:
Traceback (most recent call last):
  File "/usr/lib64/python3.9/threading.py", line 980, in _bootstrap_inner
Exception in thread data_processing:
Exception in thread data_processing:
Traceback (most recent call last):
  File "/usr/lib64/python3.9/threading.py", line 980, in _bootstrap_inner
Traceback (most recent call last):
    self.run()
  File "/usr/lib64/python3.9/threading.py", line 917, in run
  File "/usr/lib64/python3.9/threading.py", line 980, in _bootstrap_inner
    self.run()
  File "/usr/lib64/python3.9/threading.py", line 917, in run
    self._target(*self._args, **self._kwargs)
  File "/appl/AS/4.1.2/lib/python3.9/site-packages/autosubmit-4.1.2-py3.9.egg/autosubmit/job/job_packages.py", line 103, in check_scripts
    raise AutosubmitCritical(f"Job script:{job.file} does not exists",7014)
log.log.AutosubmitCritical:  
    self._target(*self._args, **self._kwargs)
  File "/appl/AS/4.1.2/lib/python3.9/site-packages/autosubmit-4.1.2-py3.9.egg/autosubmit/job/job_packages.py", line 103, in check_scripts
    self.run()
    raise AutosubmitCritical(f"Job script:{job.file} does not exists",7014)
  File "/usr/lib64/python3.9/threading.py", line 917, in run
log.log.AutosubmitCritical:  
    self._target(*self._args, **self._kwargs)
  File "/appl/AS/4.1.2/lib/python3.9/site-packages/autosubmit-4.1.2-py3.9.egg/autosubmit/job/job_packages.py", line 103, in check_scripts
    raise AutosubmitCritical(f"Job script:{job.file} does not exists",7014)
log.log.AutosubmitCritical:  
Exception in thread data_processing:
Traceback (most recent call last):
  File "/usr/lib64/python3.9/threading.py", line 980, in _bootstrap_inner
    self.run()
  File "/usr/lib64/python3.9/threading.py", line 917, in run
    self._target(*self._args, **self._kwargs)
  File "/appl/AS/4.1.2/lib/python3.9/site-packages/autosubmit-4.1.2-py3.9.egg/autosubmit/job/job_packages.py", line 113, in _create_scripts_threaded
    self._job_scripts[jobs[i].name] = jobs[i].create_script(configuration)
  File "/appl/AS/4.1.2/lib/python3.9/site-packages/autosubmit-4.1.2-py3.9.egg/autosubmit/job/job.py", line 2065, in create_script
Exception in thread data_processing:
Traceback (most recent call last):
  File "/usr/lib64/python3.9/threading.py", line 980, in _bootstrap_inner
    template_content,additional_templates = self.update_content(as_conf)
  File "/appl/AS/4.1.2/lib/python3.9/site-packages/autosubmit-4.1.2-py3.9.egg/autosubmit/job/job.py", line 1954, in update_content
    self.run()
  File "/usr/lib64/python3.9/threading.py", line 917, in run
    self.update_parameters(as_conf, self.parameters)
    self._target(*self._args, **self._kwargs)
  File "/appl/AS/4.1.2/lib/python3.9/site-packages/autosubmit-4.1.2-py3.9.egg/autosubmit/job/job.py", line 1912, in update_parameters
  File "/appl/AS/4.1.2/lib/python3.9/site-packages/autosubmit-4.1.2-py3.9.egg/autosubmit/job/job_packages.py", line 113, in _create_scripts_threaded
    self.update_dict_parameters(as_conf)
  File "/appl/AS/4.1.2/lib/python3.9/site-packages/autosubmit-4.1.2-py3.9.egg/autosubmit/job/job.py", line 1737, in update_dict_parameters
    self._job_scripts[jobs[i].name] = jobs[i].create_script(configuration)
  File "/appl/AS/4.1.2/lib/python3.9/site-packages/autosubmit-4.1.2-py3.9.egg/autosubmit/job/job.py", line 2065, in create_script
    template_content,additional_templates = self.update_content(as_conf)
  File "/appl/AS/4.1.2/lib/python3.9/site-packages/autosubmit-4.1.2-py3.9.egg/autosubmit/job/job.py", line 1954, in update_content
    self.update_parameters(as_conf, self.parameters)
    self.retrials = as_conf.jobs_data.get(self.section,{}).get("RETRIALS", as_conf.experiment_data.get("CONFIG",{}).get("RETRIALS", 0))
  File "/appl/AS/4.1.2/lib/python3.9/site-packages/autosubmit-4.1.2-py3.9.egg/autosubmit/job/job.py", line 1912, in update_parameters
  File "/appl/AS/4.1.2/lib/python3.9/site-packages/autosubmitconfigparser-1.0.58-py3.9.egg/autosubmitconfigparser/config/configcommon.py", line 78, in jobs_data
    self.update_dict_parameters(as_conf)
  File "/appl/AS/4.1.2/lib/python3.9/site-packages/autosubmit-4.1.2-py3.9.egg/autosubmit/job/job.py", line 1737, in update_dict_parameters
    return self.experiment_data["JOBS"]
KeyError: 'JOBS'
    self.retrials = as_conf.jobs_data.get(self.section,{}).get("RETRIALS", as_conf.experiment_data.get("CONFIG",{}).get("RETRIALS", 0))
  File "/appl/AS/4.1.2/lib/python3.9/site-packages/autosubmitconfigparser-1.0.58-py3.9.egg/autosubmitconfigparser/config/configcommon.py", line 78, in jobs_data
    return self.experiment_data["JOBS"]
KeyError: 'JOBS'
Exception in thread data_processing:
Traceback (most recent call last):
  File "/usr/lib64/python3.9/threading.py", line 980, in _bootstrap_inner
    self.run()
  File "/usr/lib64/python3.9/threading.py", line 917, in run
    self._target(*self._args, **self._kwargs)
  File "/appl/AS/4.1.2/lib/python3.9/site-packages/autosubmit-4.1.2-py3.9.egg/autosubmit/job/job_packages.py", line 113, in _create_scripts_threaded
    self._job_scripts[jobs[i].name] = jobs[i].create_script(configuration)
  File "/appl/AS/4.1.2/lib/python3.9/site-packages/autosubmit-4.1.2-py3.9.egg/autosubmit/job/job.py", line 2065, in create_script
    template_content,additional_templates = self.update_content(as_conf)
  File "/appl/AS/4.1.2/lib/python3.9/site-packages/autosubmit-4.1.2-py3.9.egg/autosubmit/job/job.py", line 1954, in update_content
    self.update_parameters(as_conf, self.parameters)
  File "/appl/AS/4.1.2/lib/python3.9/site-packages/autosubmit-4.1.2-py3.9.egg/autosubmit/job/job.py", line 1912, in update_parameters
    self.update_dict_parameters(as_conf)
  File "/appl/AS/4.1.2/lib/python3.9/site-packages/autosubmit-4.1.2-py3.9.egg/autosubmit/job/job.py", line 1737, in update_dict_parameters
    self.retrials = as_conf.jobs_data.get(self.section,{}).get("RETRIALS", as_conf.experiment_data.get("CONFIG",{}).get("RETRIALS", 0))
  File "/appl/AS/4.1.2/lib/python3.9/site-packages/autosubmitconfigparser-1.0.58-py3.9.egg/autosubmitconfigparser/config/configcommon.py", line 78, in jobs_data
    return self.experiment_data["JOBS"]
KeyError: 'JOBS'
[ERROR] Trace: None
[ERROR] Error in platform LUMI for section TRANSFER: 'a0yd_19910301_fc0_1115_TRANSFER' [eCode=6015]
```

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Apr 18, 2024, 08:23

v4.1.3 had some of the critical fixes to this. We did some testing through slack


v4.1.4 further fixes issues related to wrappers, mainly regarding retrials and historical data

**Comment by kinow** [comment]
In GitLab by @Lerriola on Apr 18, 2024, 09:47

Thanks @dbeltrankyl! Should I switch and use v4.1.4 from now on?

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Apr 18, 2024, 10:24

Hello, 

You can use v4.1.3 as long as you only using horizontal wrappers but I would like to double-check that the logs are working fine in v4.1.4 and the rest of the changes

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Apr 30, 2024, 17:02

This should be fixed already in versions <=4.1.5
